esx_service/vmodl/VsphereContainerServiceImpl.py

```diff
@@ -0,0 +1,248 @@
+                    tenant.vms.append(vm_name)
+
+            # Populate associated privileges
+            if t.privileges:
+                for p in t.privileges:
+                    dap = vim.vcs.storage.DatastoreAccessPrivilege()
+                    dap.datastore =  vmdk_utils.get_datastore_name(p.datastore_url)
+                    dap.allow_create = p.allow_create
+                    dap.volume_max_size = p.max_volume_size
+                    dap.volume_total_size = p.usage_quota
+                    tenant.privileges.append(dap)
+
+            result.append(tenant)
+
+        logging.info("Successfully retrieved tenant(s): name=%s", name)
+        return result
+            
+    def UpdateTenant(self, name, new_name=None, description=None, default_datastore=None):
+        logging.info("Updating tenant: name=%s, new_name=%s, description=%s, default_datastore=%s",
+            name, new_name, description, default_datastore)
+
+        error_info = auth_api._tenant_update(name, new_name, description, default_datastore)
+        if error_info == ErrorCode.TENANT_NOT_EXIST:
+            raise vim.fault.NotFound(error_info.format(name))
+
+        if error_info == ErrorCode.TENANT_ALREADY_EXIST:
+            raise vim.fault.AlreadyExists(error_info.format(new_name))
+
+        if error_info:
+            logging.error("Failed to update tenant: name=%s, new_name=%s, description=%s, default_datastore=%s",
+            name, new_name, description, default_datastore)
+            raise vim.fault.VcsFault(msg=error_info.msg)
+        
+        logging.info("Successfully updated tenant: name=%s, new_name=%s, description=%s, default_datastore=%s",
+            name, new_name, description, default_datastore)
+
+    def AddVMs(self, tenant, vms):
+        if len(vms) == 0:
+            logging.warn("Adding VMs: the VM list is empty")
+            return
+
+        logging.info("Adding VMs: %s to tenant: %s", vms, tenant.name)
+
+        error_info = auth_api._tenant_vm_add(tenant.name, vms)
+        if error_info:
+            logging.error("Failed to add VMs: %s to tenant: ", vms, tenant.name)
+            raise vim.fault.VcsFault(msg=error_info.msg)
+
+        logging.info("Succssfully added VMs: %s to tenant: %s", vms, tenant.name)
+    
+    def RemoveVMs(self, tenant, vms):
+        if len(vms) == 0:
+            logging.warn("Remove VMs: the VM list is empty")
+            return
+
+        logging.info("Removing VMs: %s from tenant: %s", vms, tenant.name)
+
+        error_info = auth_api._tenant_vm_rm(tenant.name, vms)
+        if error_info:
+            logging.error("Failed to remove VMs: %s from tenant: ", vms, tenant.name)
+            raise vim.fault.VcsFault(msg=error_info.msg)
+
+        logging.info("Succssfully removed VMs: %s from tenant: %s", vms, tenant.name)
+    
+    def ReplaceVMs(self, tenant, vms):
+        if len(vms) == 0:
+            logging.warn("Replace VMs: the VM list is empty")
```

Comment:
I guess you meant to return error for all 3 APIs: Add VMs, RemoveVMs and ReplaceVMs.

I'm fine with this, will fix.